src/storage-client/src/client.rs

```diff
@@ -423,6 +437,32 @@ where
                     Some(Ok(StorageResponse::FrontierUppers(new_uppers)))
                 }
             }
+            StorageResponse::DroppedIds(dropped_ids) => {
+                let mut new_drops = vec![];
+
+                for id in dropped_ids {
+                    let (_, shard_frontiers) = match self.uppers.get_mut(&id) {
+                        Some(value) => value,
+                        None => panic!("Reference to absent collection: {id}"),
+                    };
+                    let prev = shard_frontiers[shard_id].take();
+                    assert!(
+                        prev.is_some(),
+                        "got double drop for {id} from shard {shard_id}"
+                    );
+
+                    if shard_frontiers.iter().all(Option::is_none) {
+                        self.uppers.remove(&id);
+                        new_drops.push(id);
+                    }
```

Comment:
so this is saying that ALL storage shards have to give their `DroppedIds` response before we clear stuff out?

Reply:
Exactly